hw/misc/hermes.c

```diff
@@ -254,6 +265,113 @@ static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
 {
 }
 
+static int __do_dma(struct HermesState *hermes, hwaddr desc_addr,
+                       uint8_t num_desc, bool h2c)
+{
+    struct hermes_dma_desc *desc;
+    hwaddr src_addr, dst_addr, nxt_addr;
+    uint8_t nxt_adj;
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+
+    trace_hermes_dma_num_adj(num_desc - 1);
+    desc = g_new(struct hermes_dma_desc, num_desc);
+    if (!desc) {
+        fprintf(stderr, "[Hermes] Failed to alloc memory for DMA descriptor\n");
+        return -1;
+    }
+
+    /* Read DMA descriptors */
+    pci_dma_read(&hermes->pdev, desc_addr, desc, num_desc * sizeof(*desc));
+    for (int i = 0; i < num_desc; i++) {
+        trace_hermes_dma_desc(desc[i].ctrl, desc[i].len, desc[i].src_addr_lo,
+                              desc[i].src_addr_hi, desc[i].dst_addr_lo,
+                              desc[i].dst_addr_hi, desc[i].nxt_addr_lo,
+                              desc[i].nxt_addr_hi);
+    }
+
+    /* DMA all descriptors in this block */
+    for (int i = 0; i < num_desc; i++) {
+        src_addr = ((uint64_t) desc[i].src_addr_hi << 32) | desc[i].src_addr_lo;
+        dst_addr = ((uint64_t) desc[i].dst_addr_hi << 32) | desc[i].dst_addr_lo;
```

Comment:
Is there a reason these can't just be 64bit types?

Reply:
Nope, fixed